src/sentry/ratelimits/cardinality.py

```diff
@@ -132,3 +157,266 @@ def use_quotas(
+        return self.cluster_num_shards / self.cluster_num_physical_shards
+
+    def check_within_quotas(
+        self, requests: Sequence[RequestedQuota], timestamp: Optional[Timestamp] = None
+    ) -> Tuple[Timestamp, Sequence[GrantedQuota]]:
+        if timestamp is None:
+            timestamp = int(time.time())
+        else:
+            timestamp = int(timestamp)
+
+        unit_keys_to_get: List[str] = []
+        set_keys_to_count: List[str] = []
+
+        for request in requests:
+            if request.quotas:
+                for hash in request.unit_hashes:
+                    unit_keys_to_get.append(self._get_timeseries_key(request, hash))
+
+            for quota in request.quotas:
+                set_keys_to_count.extend(self._get_read_sets_keys(request, quota, timestamp))
+
+        if not unit_keys_to_get and not set_keys_to_count:
+            # If there are no keys to fetch (i.e. there are no quotas to
+            # enforce), we can save the redis call entirely and just grant all
+            # quotas immediately.
+            return timestamp, [
+                GrantedQuota(
+                    request=request, granted_unit_hashes=request.unit_hashes, reached_quotas=[]
+                )
+                for request in requests
+            ]
+
+        with self.client.pipeline(transaction=False) as pipeline:
+            pipeline.mget(unit_keys_to_get)
+            # O(self.cluster_shard_factor * len(requests)), assuming there's
+            # only one per-org quota
+            for key in set_keys_to_count:
+                pipeline.scard(key)
+
+            results = iter(pipeline.execute())
+            unit_keys = dict(zip(unit_keys_to_get, next(results)))
+            set_counts = dict(zip(set_keys_to_count, results))
+
+        grants = []
+        cardinality_sample_factor = self._get_set_cardinality_sample_factor()
+        for request in requests:
+            if not request.quotas:
+                grants.append(
+                    GrantedQuota(
+                        request=request, granted_unit_hashes=request.unit_hashes, reached_quotas=[]
+                    )
+                )
+                continue
+
+            granted_hashes = []
+
+            quotas_by_remaining_limit = defaultdict(list)
+            for quota in request.quotas:
+                remaining_limit = max(
+                    0,
+                    quota.limit
+                    - cardinality_sample_factor
+                    * sum(
+                        set_counts[k] for k in self._get_read_sets_keys(request, quota, timestamp)
+                    ),
+                )
+                quotas_by_remaining_limit[remaining_limit].append(quota)
+
+            # Determine the quota(s) with the smallest remaining limit
+            # (quota.limit - <set count from redis>). If we drop any hashes at
+            # all, we can only have hit the most restrictive quota, so we
+            # report those as `reached_quotas`.
```

Comment:
What if we simplified this and allowed one quota only? 
This system is so memory intensive on redis that we are very unlikely to try to add multiple quotas. 
The algorithm would be much simpler, thus it would be easier to test and troubleshoot.

Reply:
i pushed a commit that does this, but i don't think it is a big simplification to be honest